domain/agentprovisioner/service/service.go

```diff
@@ -0,0 +1,262 @@
+		Https:   modelConfig[config.HTTPSProxyKey],
+		Ftp:     modelConfig[config.FTPProxyKey],
+		NoProxy: modelConfig[config.NoProxyKey],
+	}
+	result.JujuProxy = proxy.Settings{
+		Http:    modelConfig[config.JujuHTTPProxyKey],
+		Https:   modelConfig[config.JujuHTTPSProxyKey],
+		Ftp:     modelConfig[config.JujuFTPProxyKey],
+		NoProxy: modelConfig[config.JujuNoProxyKey],
+	}
+	result.AptProxy = proxy.Settings{
+		Http:    addSchemeIfMissing("http", getWithFallback(modelConfig, config.AptHTTPProxyKey, config.JujuHTTPProxyKey, config.HTTPProxyKey)),
+		Https:   addSchemeIfMissing("https", getWithFallback(modelConfig, config.AptHTTPSProxyKey, config.JujuHTTPSProxyKey, config.HTTPSProxyKey)),
+		Ftp:     addSchemeIfMissing("ftp", getWithFallback(modelConfig, config.AptFTPProxyKey, config.JujuFTPProxyKey, config.FTPProxyKey)),
+		NoProxy: aptNoProxy(modelConfig),
+	}
+	result.AptMirror = modelConfig[config.AptMirrorKey]
+	result.SnapProxy = proxy.Settings{
+		Http:  modelConfig[config.SnapHTTPProxyKey],
+		Https: modelConfig[config.SnapHTTPSProxyKey],
+	}
+	result.SnapStoreAssertions = modelConfig[config.SnapStoreAssertionsKey]
+	result.SnapStoreProxyID = modelConfig[config.SnapStoreProxyKey]
+	result.SnapStoreProxyURL = modelConfig[config.SnapStoreProxyURLKey]
+	result.CloudInitUserData, _ = ensureStringMaps(modelConfig[config.CloudInitUserDataKey])
+	result.ContainerInheritProperties = modelConfig[config.ContainerInheritPropertiesKey]
+	return result, nil
+}
+
+// addSchemeIfMissing adds a scheme to a URL if it is missing
+// Copied from github.com/juju/juju/environs/config
+func addSchemeIfMissing(defaultScheme string, url string) string {
+	if url != "" && !strings.Contains(url, "://") {
+		url = defaultScheme + "://" + url
+	}
+	return url
+}
+
+// Copied from github.com/juju/juju/environs/config
+func getWithFallback(c map[string]string, key, fallback1, fallback2 string) string {
+	value := c[key]
+	if value == "" {
+		value = c[fallback1]
+	}
+	if value == "" {
+		value = c[fallback2]
+	}
+	return value
+}
+
+// AptNoProxy returns the 'apt-no-proxy' for the model.
+// Copied from github.com/juju/juju/environs/config
+func aptNoProxy(c map[string]string) string {
+	value := c[config.AptNoProxyKey]
+	if value == "" {
+		if hasLegacyProxy(c) {
+			value = c[config.NoProxyKey]
+		} else {
+			value = c[config.JujuNoProxyKey]
+		}
+	}
+	return value
+}
+
+// HasLegacyProxy returns true if there is any proxy set using the old legacy proxy keys.
+// Copied from github.com/juju/juju/environs/config
+func hasLegacyProxy(c map[string]string) bool {
+	// We exclude the no proxy value as it has default value.
+	return c[config.HTTPProxyKey] != "" ||
+		c[config.HTTPSProxyKey] != "" ||
+		c[config.FTPProxyKey] != ""
+}
+
+// ensureStringMaps takes in a string and returns YAML in a map
+// where all keys of any nested maps are strings.
+// Copied from github.com/juju/juju/environs/config
+func ensureStringMaps(in string) (map[string]any, error) {
```

Comment:
I don't think this is needed. We shouldn't be trying to fix a value on the way out. This should be happening on the way in. If we look at the env.Config code it does this check for us. I am pretty certain that this is getting run as part of the ModelConfig update and set code.

Are you able to confirm. This also seems like legacy from bson days. We also throw away the euro which seems pointless. Would you mind taking a look at the code paths and confirming what I have found?

Reply:
We do have to convert it from a string to map[string]any purely for type reasons, but we can just use yaml.Unmarshal for that. So we don't need this method.